src/vm/i386/umthunkstub.S

```diff
@@ -154,34 +153,48 @@ LOCAL_LABEL(UMThunkStub_CopyStackArgs):
     and     esp, -16    // align with 16 byte
     lea     edi, [esp]
 
+    // First, we copy arguments to ecx and edx registers (if needed).
     mov     edx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
     mov     edx, dword ptr [edx + UMEntryThunk__m_pUMThunkMarshInfo]
     mov     ebx, dword ptr [edx + UMThunkMarshInfo__m_ecxArgOffset]
     cmp     ebx, -1
-    je      LOCAL_LABEL(CopyStack)
+    je      LOCAL_LABEL(InitCopyStack)
     mov     ecx, dword ptr [esi + ebx]
     add     eax, -4
     jz      LOCAL_LABEL(UMThunkStub_ArgumentsSetup)
 
     mov     ebx, dword ptr [edx + UMThunkMarshInfo__m_edxArgOffset]
     cmp     ebx, -1
-    je      LOCAL_LABEL(CopyStack)
+    je      LOCAL_LABEL(InitCopyStack)
     mov     edx, dword ptr [esi + ebx]
     add     eax, -4
     jz      LOCAL_LABEL(UMThunkStub_ArgumentsSetup)
 
-
-LOCAL_LABEL(CopyStack):
+LOCAL_LABEL(InitCopyStack):
     push    ecx
-    mov     ebx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
-    mov     ebx, dword ptr [ebx + UMEntryThunk__m_pUMThunkMarshInfo]
-    mov     ebx, [ebx + UMThunkMarshInfo__m_cbActualArgSize]
-    sub     ebx, eax
+    push    edx
+    mov     eax, 0  // Offset in edi
+    mov     ebx, 0  // Offset in esi
+LOCAL_LABEL(CopyStack):
+    mov     edx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
+    mov     edx, dword ptr [edx + UMEntryThunk__m_pUMThunkMarshInfo]
+    mov     ecx, dword ptr [edx + UMThunkMarshInfo__m_ecxArgOffset]
+    cmp     ecx, ebx
+    je      LOCAL_LABEL(IncreaseOffset)
+    mov     ecx, dword ptr [edx + UMThunkMarshInfo__m_edxArgOffset]
```

Comment:
You can use directly cmp ebx, dword ptr [edx + UMThunkMarshInfo__m_edxArgOffset] instead of loading the value to ecx first and then comparing it. At all three places in this function.